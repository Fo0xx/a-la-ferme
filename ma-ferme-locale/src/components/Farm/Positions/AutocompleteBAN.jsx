
import React from "react";

import { debounce } from "lodash"
import { Autocomplete, Box, Grid, TextField, Typography } from "@mui/material";
import axios from "axios";
import LocationOnIcon from '@mui/icons-material/LocationOn';

export default function AutocompleteBAN({ maxResults, setLocation }) {

    const [value, setValue] = React.useState(null);
    const [inputValue, setInputValue] = React.useState('');
    const [options, setOptions] = React.useState([]);

    /* A function that is called when the component is mounted. It is used to fetch the data from the
    API. */
    const fetch = React.useMemo( // memoize the fetching of the data to avoid unnecessary requests
        () =>
            debounce((request, callback) => {
                axios.get('http://api-adresse.data.gouv.fr/search/?q=' + request + '&limit=' + maxResults)
                    .then(response => {
                        callback(response.data.features);
                    }
                    )
                    .catch(error => {
                        console.log(error);
                    });
            }, 500),
        [],
    );

    React.useEffect(() => {
        let active = true;

        if (inputValue === '') {
            setOptions(value ? [value] : []);
            return undefined;
        }

        fetch(inputValue, (results) => {
            if(active) {
                let newOptions = [];

                if (value) {
                    newOptions = [value];
                }

                if(results) {
                    newOptions = [...newOptions, ...results];
                }
                
                setOptions(newOptions);
            }
        });

        return () => {
            active = false;
        };

    }, [value, inputValue, fetch]);

    return (
        <Autocomplete
            id="combo-box-demo"
            options={options}
            filterOptions={(x) => x}
            autoComplete
            includeInputInList
            filterSelectedOptions
            value={value}
            getOptionLabel={option => option.properties.name + ', ' + option.properties.postcode + ' ' + option.properties.city}
            isOptionEqualToValue={(option, value) => option.properties.label === value.properties.label}
            onChange={(event, newValue) => {
                event.preventDefault();
                
                setOptions(newValue ? [newValue, ...options] : options);
                setValue(newValue);
                // Next line is needed to avoid an error when we click on the cross to delete the value
                // It is because if we click on the cross, the value is set to null and we can't read the new value
                newValue ? setLocation(newValue.geometry.coordinates[0], newValue.geometry.coordinates[1]) : setLocation(0, 0);
            }}
            onInputChange={(event, newInputValue) => {
                event.preventDefault();

                setInputValue(newInputValue);
            }}
            renderInput={params => (<TextField {...params} label="Rechercher une adresse" variant="outlined" />)}
            renderOption={(props, option) => {
                return (
                    <li {...props}>
                        <Grid container alignItems="center">
                            <Grid item>
                                <Box
                                    component={LocationOnIcon}
                                    sx={{ color: 'text.secondary', mr: 2 }}
                                />
                            </Grid>
                            <Grid item xs>
                                    <span key={option.properties.id}>
                                        {option.properties.name}
                                    </span>

                                <Typography variant="body2" color="text.secondary">
                                    {option.properties.postcode + ' ' + option.properties.city}
                                </Typography>
                            </Grid>
                        </Grid>
                    </li>
                );
            }}

        />
    );
}



