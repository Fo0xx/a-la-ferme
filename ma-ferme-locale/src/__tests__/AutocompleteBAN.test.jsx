import { cleanup, render } from '@testing-library/react';
import AutocompleteBAN from '../components/Farm/Positions/AutocompleteBAN';
import Enzyme from 'enzyme';
import { mount, shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });

describe('AutocompleteBAN components', () => {

    afterEach(cleanup);
    const onUseAutompleteMock = jest.fn();

    it('renders correctly', () => {
        expect(mount(<AutocompleteBAN />)).toMatchSnapshot();
    });

    describe('testing functionnality', () => {

        let wrapper;

        beforeEach(() => {
            wrapper = shallow(<AutocompleteBAN />);
        });

        it('should call onUseAutomplete when the user clicks on the button', () => {
            const wrapper = shallow(<AutocompleteBAN onUseAutomplete={onUseAutompleteMock} />);
            wrapper.find('button').simulate('click');
            expect(onUseAutompleteMock).toHaveBeenCalled();
        });

        it('It\'s different after click', () => {
            const { getByText } = render(<AutocompleteBAN />);
            getByText.find('input').simulate('click');
            expect(getByText).not.toMatchSnapshot();
        });

    });

});
