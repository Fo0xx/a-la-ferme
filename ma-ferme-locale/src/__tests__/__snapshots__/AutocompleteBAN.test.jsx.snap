// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocompleteBAN components renders correctly 1`] = `ReactWrapper {}`;
